File: AlgoPersonalStudy/src/backjoon/silver/BJ_1926.java
package backjoon.silver;

import java.util.*;
import java.io.*;
public class BJ_1926 {

    static int[] delta_x = {0,1,0,-1}, delta_y = {1,0,-1,0};
    public static void main(String[] args) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int[][] map = new int[n][m];

        for(int i=0;i<n;i++){
            st = new StringTokenizer(br.readLine());
            for(int j=0;j<m;j++){
                map[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        Queue<int[]> queue = new LinkedList<>();
        boolean[][] visited = new boolean[n][m];

        int count = 0, maxSize = 0;

        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(map[i][j] == 1){
                    if(!visited[i][j]) {
                        count++;
                        int tmpSize = 1;

                        visited[i][j] = true;
                        queue.add(new int[]{i,j});
                        while(!queue.isEmpty()){
                            int[] tmp = queue.poll();
                            for(int d=0;d<4;d++){
                                int dy = tmp[0] + delta_y[d];
                                int dx = tmp[1] + delta_x[d];
                                if(dy >= 0 && dy < n && dx >=0 && dx < m){
                                    if(map[dy][dx] == 1 && !visited[dy][dx]){
                                        visited[dy][dx] = true;
                                        tmpSize++;
                                        queue.add(new int[]{dy,dx});
                                    }
                                }
                            }
                        }
                        maxSize = Math.max(maxSize, tmpSize);
                    }
                }
            }
        }
        System.out.println(count);
        System.out.println(maxSize);
    }
}

File: AlgoPersonalStudy/src/_kotlin/backjoon/silver/BJ_1926.kt
package _kotlin.backjoon.silver

import java.io.*
import java.util.*

class BJ_1926 {
    val deltaX = arrayOf(0,1,0,-1)
    val deltaY = arrayOf(1,0,-1,0)
    fun solution(){
        val br = BufferedReader(InputStreamReader(System.`in`))
        val (n,m) = br.readLine().split(" ").map { it.toInt() }

        val map = Array(n){Array(m){0} }
        val visited = Array(n){Array(m){false} }

        for(y:Int in 0 until n){
            val strArr = br.readLine().split(" ").map { it.toInt() }
            for(x:Int in 0 until m){
                map[y][x] = strArr[x]
            }
        }
        val queue = LinkedList<Array<Int>>()
        var count = 0
        var maxSize = 0

        for(y:Int in 0 until n){
            for(x:Int in 0 until m){
                if(map[y][x] == 1){
                    if(!visited[y][x]){
                        count++
                        var tmpSize = 1
                        visited[y][x] = true
                        queue.add(arrayOf(y,x))
                        while(!queue.isEmpty()){
                            val tmp = queue.poll()
                            for(d:Int in 0 until 4){
                                val dy = tmp[0] + deltaY[d]
                                val dx = tmp[1] + deltaX[d]
                                if (dy in 0..<n && dx in 0 ..< m){
                                    if(map[dy][dx] == 1 && !visited[dy][dx]){
                                        visited[dy][dx] = true
                                        tmpSize++
                                        queue.add(arrayOf(dy, dx))
                                    }
                                }
                            }
                        }
                        maxSize = maxSize.coerceAtLeast(tmpSize)
                    }
                }
            }
        }
        println(count)
        println(maxSize)
    }
}
fun main(){
    val solutionClass = BJ_1926()
    solutionClass.solution()
}